File: frontend/src/lib/api/generation.ts
/**
 * 非同期生成ジョブAPI
 *
 * 顔シミュレーション生成を非同期ジョブとして実行するAPI。
 * 長時間の処理をバックグラウンドで実行し、ポーリングで結果を取得する。
 *
 * @example
 * ```typescript
 * // ジョブ作成
 * const job = await createGenerationJob({
 *   base_image: base64Current,
 *   target_image: base64Ideal,
 *   mode: 'morph',
 *   strength: 0.5,
 * });
 *
 * // 完了まで待機
 * const result = await waitForJobCompletion(job.job_id);
 * if (result.status === 'succeeded') {
 *   const imgSrc = result.result_image_url;
 * }
 * ```
 */

import { apiGet, apiPost, ApiError } from './client';
import {
  CreateGenerationJobRequest,
  GenerationJobStatus,
  GenerationResultData,
  JobStatus,
} from './types';

/**
 * デフォルトのポーリング間隔（ミリ秒）
 */
const DEFAULT_POLL_INTERVAL = 2000;

/**
 * デフォルトの最大待機時間（ミリ秒）
 */
const DEFAULT_MAX_WAIT_TIME = 300000; // 5分

/**
 * 生成ジョブを作成
 *
 * @param request - ジョブ作成リクエスト
 * @param authToken - 認証トークン（オプション）
 * @returns 作成されたジョブのステータス
 *
 * @example
 * ```typescript
 * const job = await createGenerationJob({
 *   base_image: base64Data,
 *   target_image: base64Ideal,
 *   mode: 'parts',
 *   parts: ['eyes', 'nose', 'lips'],
 *   strength: 0.7,
 * });
 * console.log('Job created:', job.job_id);
 * ```
 */
export async function createGenerationJob(
  request: CreateGenerationJobRequest,
  authToken?: string
): Promise<GenerationJobStatus> {
  const options = authToken ? { authToken } : undefined;
  return apiPost<GenerationJobStatus>(
    '/api/v1/simulations/generate',
    request,
    options
  );
}

/**
 * ジョブのステータスを取得
 *
 * @param jobId - ジョブID
 * @returns ジョブステータス
 */
export async function getJobStatus(jobId: string): Promise<GenerationJobStatus> {
  return apiGet<GenerationJobStatus>(`/api/v1/simulations/generate/${jobId}`);
}

/**
 * ジョブの結果を取得
 *
 * @param jobId - ジョブID
 * @returns 生成結果（Base64画像データ含む）
 * @throws ApiError - ジョブが完了していない場合
 */
export async function getJobResult(jobId: string): Promise<GenerationResultData> {
  return apiGet<GenerationResultData>(`/api/v1/simulations/generate/${jobId}/result`);
}

/**
 * ジョブが完了するまで待機
 *
 * @param jobId - ジョブID
 * @param options - 待機オプション
 * @returns 完了時のジョブステータス
 * @throws Error - タイムアウトまたはジョブ失敗時
 *
 * @example
 * ```typescript
 * const result = await waitForJobCompletion(job.job_id, {
 *   pollInterval: 1000,
 *   maxWaitTime: 60000,
 *   onProgress: (status) => {
 *     console.log(`Progress: ${status.progress}%`);
 *   },
 * });
 * ```
 */
export async function waitForJobCompletion(
  jobId: string,
  options: {
    pollInterval?: number;
    maxWaitTime?: number;
    onProgress?: (status: GenerationJobStatus) => void;
  } = {}
): Promise<GenerationJobStatus> {
  const {
    pollInterval = DEFAULT_POLL_INTERVAL,
    maxWaitTime = DEFAULT_MAX_WAIT_TIME,
    onProgress,
  } = options;

  const startTime = Date.now();

  while (true) {
    const status = await getJobStatus(jobId);

    // プログレスコールバック
    if (onProgress) {
      onProgress(status);
    }

    // 終了状態をチェック
    if (status.status === 'succeeded') {
      return status;
    }

    if (status.status === 'failed') {
      throw new Error(status.error || 'Job failed');
    }

    // タイムアウトチェック
    if (Date.now() - startTime > maxWaitTime) {
      throw new Error('Job timed out');
    }

    // 次のポーリングまで待機
    await new Promise((resolve) => setTimeout(resolve, pollInterval));
  }
}

/**
 * ジョブを作成し完了まで待機するユーティリティ
 *
 * createGenerationJob + waitForJobCompletion を組み合わせた便利関数。
 *
 * @param request - ジョブ作成リクエスト
 * @param options - 待機オプション
 * @returns 生成結果
 *
 * @example
 * ```typescript
 * try {
 *   const result = await generateAndWait({
 *     base_image: base64Current,
 *     target_image: base64Ideal,
 *     mode: 'morph',
 *     strength: 0.5,
 *   }, {
 *     onProgress: (status) => setProgress(status.progress),
 *   });
 *   setResultImage(result.result_image_url);
 * } catch (error) {
 *   console.error('Generation failed:', error);
 * }
 * ```
 */
export async function generateAndWait(
  request: CreateGenerationJobRequest,
  options: {
    authToken?: string;
    pollInterval?: number;
    maxWaitTime?: number;
    onProgress?: (status: GenerationJobStatus) => void;
  } = {}
): Promise<GenerationJobStatus> {
  const { authToken, ...waitOptions } = options;

  // ジョブ作成
  const job = await createGenerationJob(request, authToken);

  // 初期プログレスコールバック
  if (waitOptions.onProgress) {
    waitOptions.onProgress(job);
  }

  // 完了まで待機
  return waitForJobCompletion(job.job_id, waitOptions);
}

/**
 * Base64画像をFileに変換するユーティリティ
 *
 * @param base64 - Base64エンコードされた画像データ
 * @param filename - ファイル名
 * @returns Fileオブジェクト
 */
export function base64ToFile(base64: string, filename: string = 'image.png'): File {
  // Data URL形式の場合、プレフィックスを除去
  const base64Data = base64.startsWith('data:')
    ? base64.split(',')[1]
    : base64;

  const byteCharacters = atob(base64Data);
  const byteNumbers = new Array(byteCharacters.length);
  for (let i = 0; i < byteCharacters.length; i++) {
    byteNumbers[i] = byteCharacters.charCodeAt(i);
  }
  const byteArray = new Uint8Array(byteNumbers);

  return new File([byteArray], filename, { type: 'image/png' });
}

/**
 * FileをBase64に変換するユーティリティ
 *
 * @param file - Fileオブジェクト
 * @returns Base64エンコードされた文字列（Data URLプレフィックス付き）
 */
export function fileToBase64(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => resolve(reader.result as string);
    reader.onerror = reject;
    reader.readAsDataURL(file);
  });
}

File: backend/app/routers/generation.py
"""Generation Jobs API router - async face generation."""

from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Header, HTTPException, Request
from fastapi.responses import JSONResponse

from app.models.schemas import (
    CreateGenerationJobRequest,
    CreateGenerationJobResponse,
    ErrorCodes,
    ErrorDetail,
    ErrorResponse,
    GenerationJobStatus,
    GenerationJobStatusResponse,
    GenerationResultData,
    GenerationResultResponse,
)
from app.services.job_queue import JobNotFoundError, get_job_queue_service

router = APIRouter(prefix="/simulations/generate", tags=["generation"])


def _parse_datetime(value: Optional[str]) -> Optional[datetime]:
    """Parse ISO datetime string, handling timezone suffix."""
    if not value:
        return None
    try:
        # Handle 'Z' suffix and '+00:00'
        if value.endswith("Z"):
            value = value[:-1] + "+00:00"
        return datetime.fromisoformat(value)
    except (ValueError, TypeError):
        return None


def _job_to_status(job: dict) -> GenerationJobStatus:
    """Convert job record to status response."""
    # For result URL, use the stored base64 directly
    result_url = None
    if job.get("result_image_path") and job.get("status") == "succeeded":
        # Return as data URL for now (or storage URL in future)
        result_path = job["result_image_path"]
        if result_path.startswith("data:") or len(result_path) > 200:
            # It's base64 data, return as-is or as data URL
            result_url = result_path if result_path.startswith("data:") else f"data:image/png;base64,{result_path}"
        else:
            result_url = result_path  # Storage URL

    return GenerationJobStatus(
        job_id=job["id"],
        status=job["status"],
        progress=job.get("progress", 0),
        result_image_url=result_url,
        error=job.get("error"),
        created_at=_parse_datetime(job.get("created_at")) or datetime.utcnow(),
        started_at=_parse_datetime(job.get("started_at")),
        completed_at=_parse_datetime(job.get("completed_at")),
    )


def _verify_optional_token(authorization: Optional[str] = Header(None)) -> Optional[str]:
    """Optionally verify JWT token and return user_id."""
    if not authorization or not authorization.startswith("Bearer "):
        return None

    token = authorization.replace("Bearer ", "")

    try:
        import jwt

        decoded = jwt.decode(token, options={"verify_signature": False})
        return decoded.get("sub")
    except Exception:
        return None


@router.post("", response_model=CreateGenerationJobResponse)
async def create_generation_job(
    request: Request,
    data: CreateGenerationJobRequest,
    authorization: Optional[str] = Header(None),
):
    """
    Create a new face generation job.

    The job will be queued and processed asynchronously by the worker.
    Poll the status endpoint to check progress and get results.

    - **base_image**: Base64 encoded base face image (the person)
    - **target_image**: Base64 encoded target/ideal face image (features to apply)
    - **mode**: 'morph' for full face blending, 'parts' for selective features
    - **parts**: For mode='parts', list of parts to blend (eyes, nose, lips, etc.)
    - **strength**: How much to blend toward target (0-1, default 0.5)
    """
    try:
        # Validate parts for parts mode
        if data.mode == "parts":
            if not data.parts or len(data.parts) == 0:
                return JSONResponse(
                    status_code=400,
                    content=ErrorResponse(
                        error=ErrorDetail(
                            code=ErrorCodes.VALIDATION_ERROR,
                            message="Parts list is required for mode='parts'",
                        )
                    ).model_dump(),
                )

        # Get optional user_id
        user_id = _verify_optional_token(authorization)

        # Create job
        job_service = get_job_queue_service()
        job = job_service.create_job(
            mode=data.mode,
            base_image_path=data.base_image,
            target_image_path=data.target_image,
            parts=data.parts,
            strength=data.strength or 0.5,
            seed=data.seed,
            user_id=user_id,
        )

        return CreateGenerationJobResponse(data=_job_to_status(job))

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.INTERNAL_ERROR,
                    message=f"Failed to create job: {str(e)}",
                )
            ).model_dump(),
        )


@router.get("/{job_id}", response_model=GenerationJobStatusResponse)
async def get_job_status(
    request: Request,
    job_id: str,
):
    """
    Get the status of a generation job.

    Returns current status, progress percentage, and result URL when complete.

    Status values:
    - **queued**: Waiting to be processed
    - **running**: Currently being processed
    - **succeeded**: Complete, result available
    - **failed**: Processing failed, see error field
    """
    try:
        job_service = get_job_queue_service()
        job = job_service.get_job(job_id)

        return GenerationJobStatusResponse(data=_job_to_status(job))

    except JobNotFoundError:
        return JSONResponse(
            status_code=404,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.NOT_FOUND,
                    message=f"Job {job_id} not found",
                )
            ).model_dump(),
        )
    except Exception as e:
        return JSONResponse(
            status_code=500,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.INTERNAL_ERROR,
                    message=f"Failed to get job status: {str(e)}",
                )
            ).model_dump(),
        )


@router.get("/{job_id}/result", response_model=GenerationResultResponse)
async def get_job_result(
    request: Request,
    job_id: str,
):
    """
    Get the result of a completed generation job.

    Returns the generated image as base64 encoded data.
    Only available when job status is 'succeeded'.
    """
    try:
        job_service = get_job_queue_service()
        job = job_service.get_job(job_id)

        if job["status"] != "succeeded":
            return JSONResponse(
                status_code=400,
                content=ErrorResponse(
                    error=ErrorDetail(
                        code=ErrorCodes.JOB_FAILED,
                        message=f"Job is not complete. Status: {job['status']}",
                        details={"status": job["status"], "error": job.get("error")},
                    )
                ).model_dump(),
            )

        result_path = job.get("result_image_path")
        if not result_path:
            return JSONResponse(
                status_code=500,
                content=ErrorResponse(
                    error=ErrorDetail(
                        code=ErrorCodes.INTERNAL_ERROR,
                        message="Result image not found",
                    )
                ).model_dump(),
            )

        # Extract base64 data (remove data URL prefix if present)
        image_data = result_path
        if image_data.startswith("data:"):
            image_data = image_data.split(",", 1)[1]

        return GenerationResultResponse(
            data=GenerationResultData(
                job_id=job_id,
                image=image_data,
                format="png",
                mode=job["mode"],
                strength=job.get("strength", 0.5),
            )
        )

    except JobNotFoundError:
        return JSONResponse(
            status_code=404,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.NOT_FOUND,
                    message=f"Job {job_id} not found",
                )
            ).model_dump(),
        )
    except Exception as e:
        return JSONResponse(
            status_code=500,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.INTERNAL_ERROR,
                    message=f"Failed to get result: {str(e)}",
                )
            ).model_dump(),
        )
